### PR TITLE
SAA-1277: Remove transferred prisoners from appointments

### PR DESCRIPTION
**Pull request by iterative-it**

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentAttendeeService.kt [hunk 1]
```diff
@@ -4,57 +4,116 @@ import org.slf4j.LoggerFactory
...
-                ),
-              )
+  fun removePrisonerFromFutureAppointments(prisonCode: String, prisonerNumber: String, removedTime: LocalDateTime, removalReasonId: Long, removedBy: String) {
+    val removalReason = appointmentAttendeeRemovalReasonRepository.findOrThrowNotFound(removalReasonId)
+    appointmentInstanceRepository.findByPrisonCodeAndPrisonerNumberFromNow(prisonCode, prisonerNumber)
+      .forEach {
+        appointmentAttendeeRepository.findById(it.appointmentAttendeeId)
+          .ifPresent { attendee ->
+            transactionHandler.newSpringTransaction {
+              attendee.remove(removedTime, removalReason, removedBy)
```
Comment (iterative-it) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentAttendeeService.kt:50]: Replaced the hard delete of attendees and appointments with a soft delete of the attendees only. The prior logic path for individual appointments is no longer valid as all appointments are group appointments now

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentAttendeeService.kt [hunk 2]
```diff
@@ -4,57 +4,116 @@ import org.slf4j.LoggerFactory
...
             }
-        }
+
+            log.info("Removed appointment attendee with id '${it.appointmentAttendeeId}' for prisoner '$prisonerNumber' from appointment with id '${it.appointmentId}'")
+
+            // TODO: Rename to more generic PrisonerRemovedFromFutureAppointments and pass in reason
+            auditService.logEvent(
```
Comment (iterative-it) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentAttendeeService.kt:56]: Will raise a separate ticket to improve auditing on appointments

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentAttendeeService.kt [hunk 3]
```diff
@@ -4,57 +4,116 @@ import org.slf4j.LoggerFactory
...
+                prisonCode = it.prisonCode,
+                prisonerNumber = it.prisonerNumber,
+                createdAt = removedTime,
+              ),
+            )
+          }
+      }
+  }
+
+  fun manageAppointmentAttendees(prisonCode: String, daysAfterNow: Long) {
```
Comment (iterative-it) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentAttendeeService.kt:69]: This function was moved from the now deleted ManageAppointmentService as it is managing attendees and uses the removePrisonerFromFutureAppointments function in this service

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentAttendeeService.kt [hunk 4]
```diff
@@ -4,57 +4,116 @@ import org.slf4j.LoggerFactory
...
+
+    permanentlyReleasedPrisoners.forEach {
+      removePrisonerFromFutureAppointments(prisonCode, it.prisonerNumber, removedTime, PRISONER_STATUS_RELEASED_APPOINTMENT_ATTENDEE_REMOVAL_REASON_ID, removedBy)
+      log.info("Removed prisoner '${it.prisonerNumber}' from future appointments as they have been released")
+    }
+
+    val prisonersNotInExpectedPrison = prisoners.notInExpectedPrison(prisonCode)
+    log.info("Found ${prisonersNotInExpectedPrison.size} prisoners not in prison code '$prisonCode'")
+    val expiredMoves = prisonersNotInExpectedPrison.getExpiredMoves(regime)
+    log.info("Found ${expiredMoves.size} prisoners that left prison code '$prisonCode' more than ${regime.maxDaysToExpiry} day(s) ago")
```
Comment (iterative-it) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentAttendeeService.kt:96]: Primary change. Adds support for detecting if a prisoner left the prison more than 5 (by default) days ago after which they are considered permanently transferred

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentAttendeeService.kt [hunk 5]
```diff
@@ -4,57 +4,116 @@ import org.slf4j.LoggerFactory
...
+
+  private fun getPrisonNumbersForFutureAppointments(prisonCode: String, daysAfterNow: Long) =
+    LocalDateRange(LocalDate.now(), LocalDate.now().plusDays(daysAfterNow)).flatMap { date ->
+      appointmentRepository.findAllByPrisonCodeAndStartDate(prisonCode, date).flatMap { it.prisonerNumbers() }
+    }.distinct()
+
+  private fun List<Prisoner>.permanentlyReleased() = filter { it.isPermanentlyReleased() }
+
+  private fun List<Prisoner>.notInExpectedPrison(prisonCode: String) =
+    filterNot { it.isPermanentlyReleased() }
```
Comment (iterative-it) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentAttendeeService.kt:112]: It is possible for a prisoner and their movements to pass the logic checks for both being released and having been transferred permanently. This extra filter prevents released prisoners also being removed from future appointments using the permanent transfer reason

File: helm_deploy/hmpps-activities-management-api/templates/manage-appointment-attendees-cronjob.yaml [hunk 6]
```diff
@@ -22,4 +22,4 @@ spec:
               args:
                 - /bin/sh
                 - -c
-                - curl --fail --retry 2 -X POST "http://hmpps-activities-management-api/job/appointments/manage-attendees?daysBeforeNow=0&daysAfterNow=14"
+                - curl --fail --retry 2 -X POST "http://hmpps-activities-management-api/job/appointments/manage-attendees?daysAfterNow=14"
```
Comment (iterative-it) [inline helm_deploy/hmpps-activities-management-api/templates/manage-appointment-attendees-cronjob.yaml:25]: The job now only processes future appointments so no need for the days before now parameter

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/AppointmentJobIntegrationTest.kt [hunk 7]
```diff
@@ -33,78 +41,185 @@ import java.time.LocalDate
...
+  private val nonExpiredMovement = movement(prisonerNumber = prisonNumber, fromPrisonCode = risleyPrisonCode, movementDate = TimeSource.daysInPast(4))
+
+  @Sql("classpath:test_data/seed-manage-appointments-job.sql")
+  @Test
+  fun `do not remove released prisoner from future appointments when days from now does not include their appointments`() {
+    prisonerSearchApiMockServer.stubSearchByPrisonerNumbers(listOf("B2345CD"), listOf(activeInPrisoner))
+
+    webTestClient.manageAppointmentAttendees(0)
+
+    with(webTestClient.getAppointmentSeriesById(1)!!) {
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/AppointmentJobIntegrationTest.kt:100]: Do you need to do these two appointments assertions before the call to `webTestClient.manageAppointmentAttendees(0)` as well?

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/AppointmentJobIntegrationTest.kt [hunk 8]
```diff
@@ -33,78 +41,185 @@ import java.time.LocalDate
...
   @Test
-  fun `remove attendees for released prisoners`() {
-    with(appointmentAttendeeRepository.findAll()) {
-      size isEqualTo 18
-      onEach { it.isRemoved() isBool false }
-      onEach { it.isDeleted isBool false }
+  fun `remove released prisoner from future appointments`() {
+    prisonerSearchApiMockServer.stubSearchByPrisonerNumbers(listOf(prisonNumber, "B2345CD"), listOf(prisonerReleasedToday, activeInPrisoner))
+
+    webTestClient.manageAppointmentAttendees(1)
```
Comment (matt-crowson-moj) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/integration/AppointmentJobIntegrationTest.kt:117]: Similar comment here. An assertion to say what the number should be before?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentAttendeeService.kt [hunk 9]
```diff
@@ -4,57 +4,116 @@ import org.slf4j.LoggerFactory
...
+            )
+          }
+      }
+  }
+
+  fun manageAppointmentAttendees(prisonCode: String, daysAfterNow: Long) {
+    val removedTime = LocalDateTime.now()
+    val removedBy = "MANAGE_APPOINTMENT_SERVICE"
+
+    require(daysAfterNow in 0..60) {
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/service/AppointmentAttendeeService.kt:73]: Should that be 1..60?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/resource/AppointmentJobController.kt [hunk 10]
```diff
@@ -35,14 +35,11 @@ class AppointmentJobController(
   @ResponseBody
   @ResponseStatus(HttpStatus.ACCEPTED)
   fun triggerManageAllocationsJob(
-    @RequestParam(value = "daysBeforeNow", required = true)
-    @Parameter(description = "The number of days prior to today to start the date range. The attendees for appointments starting after this will be managed.")
-    daysBeforeNow: Long,
     @RequestParam(value = "daysAfterNow", required = true)
-    @Parameter(description = "The number of days after today to end the date range. The attendees for appointments starting before this will be managed.")
+    @Parameter(description = "The number of days into the future to manage appointments. The attendees for future appointments starting on those days this will be managed.")
```
Comment (matt-crowson-moj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsactivitiesmanagementapi/resource/AppointmentJobController.kt:39]: Perhaps include the min-max days in the doc comments?

**Review by matt-crowson-moj — DISMISSED** [review]
LGTM - a few minor questions.

**Review by timharrison-moj — DISMISSED** [review]
LGTM

**Review by matt-crowson-moj — APPROVED** [review]
LGTM

**Pull request merged**